zio-http/src/main/scala/zio/http/QueryParams.scala

```diff
@@ -64,19 +64,29 @@ case class QueryParams private[http] (map: Map[String, Chunk[String]])
 
   def encode(charset: Charset): String = QueryParamEncoding.default.encode("", self, charset)
 
+  override def equals(that: Any): Boolean = that match {
+    case that: QueryParams => self.normalize.map == that.normalize.map
+    case _                 => false
+  }
+
   override def filter(p: ((String, Chunk[String])) => Boolean): QueryParams =
     QueryParams(map.filter(p))
 
-  override def isEmpty: Boolean = map.isEmpty
-
-  def toForm: Form = Form.fromQueryParams(self)
+  override def hashCode: Int = normalize.map.hashCode
 
-  def toMap: Map[String, Chunk[String]] = map
+  override def isEmpty: Boolean = map.isEmpty
 
   override def get(key: String): Option[Chunk[String]] = map.get(key)
 
   override def iterator: Iterator[(String, Chunk[String])] = map.iterator
 
+  def normalize: QueryParams =
+    if (isEmpty) self
+    else QueryParams(map.filter(i => i._1.nonEmpty && i._2.nonEmpty))
+
+  def toForm: Form = Form.fromQueryParams(self)
+
+  def toMap: Map[String, Chunk[String]] = map
```

Comment:
Looks a bit confusing that

- we are extending `Map[String, Chunk[String]]`
- also have a public field `map: Map[String, Chunk[String]]`
- and a `toMap: Map[String, Chunk[String]]` 

and `self` is not the same map as the "inner map"; one could ask what is the difference (apart from having extra methods on `QueryParams`), are all map operations equivalent to regardless when you call `toMap`, etc.

Reply:
Agreed, will clean up 👍